From d850a8a22cae90b764ede6f251a6fb57ef025b80 Mon Sep 17 00:00:00 2001
From: John Greene <jogreene@redhat.com>
Date: Thu, 14 Aug 2014 14:42:05 -0400
Subject: [ethernet] e1000e: Fix ethtool offline tests for 82579 parts

Message-id: <1408027348-9610-12-git-send-email-jogreene@redhat.com>
Patchwork-id: 87712
O-Subject: [RHEL7.1 PATCH 11/34] e1000e: Fix ethtool offline tests for 82579 parts
Bugzilla: 1091119
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

Changes to the rar_entry_count value require a change to the indexing
used to access the SHRA[H|L] registers when testing them with
'ethtool -t <iface> offline'

Signed-off-by: Dave Ertman <davidx.m.ertman@intel.com>
Tested-by: Aaron Brown <aaron.f.brown@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
(cherry picked from commit ad40064e88df1a95a3532a35071e46d8db1fbe74)

BZ: 1091119
Testing: Acceptable, see BZ
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7834849
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/e1000e/ethtool.c b/drivers/net/ethernet/intel/e1000e/ethtool.c
index 7a47902..3c2898d 100644
--- a/drivers/net/ethernet/intel/e1000e/ethtool.c
+++ b/drivers/net/ethernet/intel/e1000e/ethtool.c
@@ -917,15 +917,21 @@ static int e1000_reg_test(struct e1000_adapter *adapter, u64 *data)
   }
   if (mac->type == e1000_pch2lan) {
    /* SHRAH[0,1,2] different than previous */
-   if (i == 7)
+   if (i == 1)
     mask &= 0xFFF4FFFF;
    /* SHRAH[3] different than SHRAH[0,1,2] */
-   if (i == 10)
+   if (i == 4)
     mask |= (1 << 30);
+   /* RAR[1-6] owned by management engine - skipping */
+   if (i > 0)
+    i += 6;
   }
 
   REG_PATTERN_TEST_ARRAY(E1000_RA, ((i << 1) + 1), mask,
            0xFFFFFFFF);
+  /* reset index to actual value */
+  if ((mac->type == e1000_pch2lan) && (i > 6))
+   i -= 6;
  }
 
  for (i = 0; i < mac->mta_reg_count; i++)
-- 
1.7.1